[FLOC-3234] Refactor IDeploy to enable state transfer from discover_state to calculate_changes.

sarum90 opened the pull request:
This refactors IDeploy to enable state to go from discover_state to calculate_changes without requiring that the state be transferred to the control service.

It does so by introducing a new interface ILocalState. discover_state is modified to return a provider of this interface. This interface has a shared_state_changes() method that returns the tuple of IClusterStateChanges that discover_state would have returned prior to this PR. The entire opaque ILocalState is also handed to calculate_changes in addition to being used to send updates to the control service.

Implementations of IDeployer can use this to transfer data from discover_state to calculate_changes, and I've added BlockDeviceDeployerLocalState as an example of how the information from api.list_volumes() can be transferred (required for upcoming changes to reduce API calls).

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/clusterhq/flocker/2046)
<!-- Reviewable:end -->

itamarst commented on flocker/node/_deploy.py:
```diff
@@ -51,6 +52,55 @@ def _to_volume_name(dataset_id):
+@implementer(ILocalState)
+class FullySharedLocalState(PClass):
+    """
+    An ``ILocalState`` provider that
```
Unfinished state.

itamarst commented on flocker/node/_deploy.py:
```diff
@@ -51,6 +52,55 @@ def _to_volume_name(dataset_id):
+
+    def shared_state_changes(self):
+        """
+        All state is shared in this implementation of LocalState.
```
ILocalState.

itamarst commented on flocker/node/_deploy.py:
```diff
@@ -75,14 +125,13 @@ def discover_state(local_state):
-            service (see ``flocker.control._protocol``) and may also be
-            passed to this object's ``calculate_changes()`` method.
+        :return: A ``Deferred`` which fires with a ``ILocalState``. The
+            state_changes will be passed to the control service (see
```
Instead of "the state_changes" maybe "the result of shared_state_changes()".

itamarst commented on flocker/node/agents/blockdevice.py:
```diff
@@ -1540,21 +1569,22 @@ def is_existing_block_device(dataset_id, path):
             ),
-            NonManifestDatasets(datasets=nonmanifest),
+            nonmanifest_datasets=NonManifestDatasets(datasets=nonmanifest),
+            volumes=[],
```
Any reason you're not filling in the volumes here?

itamarst commented on flocker/node/test/test_deploy.py:
```diff
@@ -659,18 +748,8 @@ def test_discover_one(self):
         with a a list of running ``Application``\ s; one for each active
         container.
         """
-        fake_docker = FakeDockerClient(units={APP_NAME: UNIT_FOR_APP})
```
:heart_eyes:

sarum90 replied:
:grin: yea I think I might want to do something similar with a few other chunks of code in here... this seemed like a good start.

itamarst commented on flocker/node/agents/test/test_blockdevice.py:
```diff
@@ -698,6 +736,17 @@ def add_application_with_volume(node_state):
+    Wrapper for deployer.calculate_changes with sane defaults for testing.
+    """
+    if local_state is None:
+        local_state = EMPTY_LOCAL_STATE
```
Doesn't seem like this is the right default, especially once we have code that actually uses it. Perhaps as a minimum it should be extracting info from cluster_state?

itamarst commented on flocker/node/functional/test_deploy.py:
```diff
@@ -47,30 +47,37 @@ def __init__(self, hostname, volume_service, docker_client=None,
-            manifestations_state = manifestations_state[0]
+        def got_manifestations_state(manifestations_local_state):
+            manifestations_state = (
+                manifestations_local_state.shared_state_changes()[0])
```
Given `manifestations_local_state` has an explicit attribute with the object we want, seems like that's better than using a hard to understand and brittle index into a tuple.

itamarst commented on flocker/node/functional/test_deploy.py:
```diff
@@ -47,30 +47,37 @@ def __init__(self, hostname, volume_service, docker_client=None,
+                app_state = app_local_state.shared_state_changes()
+                new_app_local_state = app_local_state.set(
+                    "cluster_state_changes",
+                    [app_state[0].evolver()
```
Ditto about using attribute instead.

itamarst commented on flocker/node/test/test_deploy.py:
```diff
@@ -126,6 +132,71 @@ def assert_application_calculated_changes(
+        shared_state_changes() is called.
+        """
+        cluster_state_changes = ClusterStateChangePVector.create([
+            ARBITRARY_CLUSTER_STATE_CHANGE, ARBITRARY_CLUSTER_STATE_CHANGE])
```
:question: maybe two different kinds of objects here?

itamarst replied:
(:question: is our shorthand for "optional")

itamarst commented on flocker/node/test/test_deploy.py:
```diff
@@ -626,32 +697,50 @@ class ApplicationNodeDeployerDiscoverNodeConfigurationTests(
         """
-        ``ApplicationNodeDeployer.discover_state`` returns an empty
-        ``NodeState`` if there are no Docker containers on the host.
+        Utility method for common code across the test cases in this class.
```
The fact it's utility can be deduced from its name and location; the interesting thing is what it actually does.

itamarst commented on flocker/node/test/test_deploy.py:
```diff
@@ -1512,7 +1455,8 @@ def test_open_port_needs_creating(self):
         result = api.calculate_changes(
             desired_configuration=desired,
-            current_cluster_state=DeploymentState(nodes=[node_state]))
+            current_cluster_state=DeploymentState(nodes=[node_state]),
```
Here an in later cases where cluster state is not empty, local state being EMPTY_LOCAL_STATE is often incorrect and not how real code would work. Specifically the cases where the node for whom state is known is the on the Deployer is for.

itamarst commented on flocker/node/testtools.py:
```diff
@@ -150,9 +152,10 @@ class DummyDeployer(object):
     node_uuid = uuid4()
 
     def discover_state(self, node_stat):
```
Pre-existing bug, but that's missing a "e" at the end.

itamarst commented on flocker/node/_deploy.py:
```diff
@@ -51,6 +52,55 @@ def _to_volume_name(dataset_id):
+        Calculate the part of the local state that needs to be sent to the
+        control service.
+
+        :return: A tuple of ``IClusterStateChange`` providers that describe
```
Requiring a tuple seems over-specific? Any sort of iterable ought to be OK.

itamarst commented on flocker/node/testtools.py:
```diff
@@ -225,13 +231,15 @@ def _discover_state(self):
+            def discovered(local_state):
+                self.assertTrue(ILocalState.providedBy(local_state))
+                self.assertEqual(tuple,
+                                 type(local_state.shared_state_changes()))
```
The tuple assert seems unnecessary given next test.

itamarst commented on flocker/node/testtools.py:
```diff
@@ -298,7 +308,12 @@ def assert_calculated_changes_for_deployer(
     :param expected_changes: The ``IStateChange`` expected to be returned.
     :param Leases leases: Currently configured leases. By default none exist.
+    :param ILocalState local_state: The local_state to pass into
+        calculate_changes.  If None, an empty ``FullySharedLocalState`` will be
```
Not a great a default, as mentioned above. Use `FullySharedLocalState(state_changes=[node_state])` or whatever it is as the default I think?

itamarst commented:
Thanks for working on this, it's going to give us a lot more flexibility compared to our previous need to add every discovered bit of information to the global state in the control service. Please address comments above and resubmit.

sarum90 commented:
Review status: 0 of 9 files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---

<sup>**[flocker/node/_deploy.py, line 67 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnCwSm_DOGOIVNO-r1-67)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/_deploy.py#L67)):</sup>
It has to be an iterable that works with AMP, I think. For instance, http://build.clusterhq.com/builders/flocker-centos-7/builds/7434 is what I hit when I just returned my ClusterStateChangesPVector.

I did not know how to test for "iterable that works with AMP", but I did know how to test for "tuple", so I figured a narrower,  tested API was better than an untested broader one. Certainly open to suggestions here.
<!-- Reviewable comment -K0CLzpe_-FTeA1hAC8h:-1534555131 -->

---

<sup>**[flocker/node/_deploy.py, line 88 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnCwSm_DOGOIVNO-r1-88)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/_deploy.py#L88)):</sup>
Done.
<!-- Reviewable comment -K0COPmYj1rSqeNwcpXH:-499146009 -->

---

<sup>**[flocker/node/_deploy.py, line 99 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnCwSm_DOGOIVNO-r1-99)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/_deploy.py#L99)):</sup>
Done.
<!-- Reviewable comment -K0COQjy_-heuhPx942C:-499146009 -->

---

<sup>**[flocker/node/_deploy.py, line 129 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnCwSm_DOGOIVNO-r1-129)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/_deploy.py#L129)):</sup>
Done.
<!-- Reviewable comment -K0CORcn_LWmrMKavwaD:-499146009 -->

---

<sup>**[flocker/node/agents/blockdevice.py, line 1584 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnCwSm_DOGOIVNQ-r1-1584)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/agents/blockdevice.py#L1584)):</sup>
Not a good one. It's a nice bite sized commit to change the tests and actually pass volumes. I'll add it to this PR rather than waiting for the next one... Done.
<!-- Reviewable comment -K0CUT7T3u2zMAIUxFQF:939432934 -->

---

<sup>**[flocker/node/agents/test/test_blockdevice.py, line 745 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnCwSm_DOGOIVNR-r1-745)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/agents/test/test_blockdevice.py#L745)):</sup>
Yea... tbh this is the one spot where I was thinking ILocalState might actually need some re-design to have 2 methods, shared_state_changes() and... [terrible_name] state_for_calculate_changes() or something. Because in a lot of the testing callsites we don't necessarily have easy access to the IClusterStateChanges needed to reconstruct an ILocalState. In fact, normally we just have a cluster_state without any IClusterStateChanges. I'll see what I can do. \me goes and implements.

Took awhile, but I think I found a solution.

We have some tests that manually construct ClusterState and then pass that into calculate_changes. The BlockDeviceDeployer also still has a reference to a valid API, but the API and backing store potentially intentionally out of sync with what we are passing in(perhaps to ensure that calculate_changes doesn't cheat and re-do API calls). That made this a little tricky, because I have to make up blockdevice_ids now since the API is not in on the joke.
<!-- Reviewable comment -K0CXbqguQlVzwFdu10s:138897943 -->

---

<sup>**[flocker/node/functional/test_deploy.py, line 52 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnCwSm_DOGOIVNS-r1-52)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/functional/test_deploy.py#L52)):</sup>
Originally these were actually FullySharedLocalState instances, but I realized that was not as clean -- Done.
<!-- Reviewable comment -K0DY5N9w69TgU7RFDjJ:325159177 -->

---

<sup>**[flocker/node/functional/test_deploy.py, line 60 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnCwSm_DOGOIVNS-r1-60)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/functional/test_deploy.py#L60)):</sup>
Done.
<!-- Reviewable comment -K0DYC3x1xinE1JPRqp9:-499146009 -->

---

<sup>**[flocker/node/test/test_deploy.py, line 193 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnDICX6If6RXNGz-r1-193)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/test/test_deploy.py#L193)):</sup>
Refactored away.
<!-- Reviewable comment -K0DYF2U8C9duAB6fOJd:1926711844 -->

---

<sup>**[flocker/node/test/test_deploy.py, line 714 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnDICX6If6RXNGz-r1-714)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/test/test_deploy.py#L714)):</sup>
Done.
<!-- Reviewable comment -K0DaMH7MOYqcg8uunkg:-499146009 -->

---

<sup>**[flocker/node/test/test_deploy.py, line 1458 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnDICX6If6RXNGz-r1-1458)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/test/test_deploy.py#L1458)):</sup>
Got rid of EMPTY_LOCAL_STATE, replaces all usages with a NodeLocalState(node_state=...known state) or with a new utility that constructs a NodeLocalState with a NodeState that only has hostname and uuid filled in (taken from the deployer) leaving the rest as None to signify unknown knowledge of the Node.
<!-- Reviewable comment -K0DoTPLqtyERt5UX-P8:1488201342 -->

---

<sup>**[flocker/node/testtools.py, line 154 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnDICX6If6RXNH--r1-154)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/testtools.py#L154)):</sup>
Done.
<!-- Reviewable comment -K0COuzWpqZ2Evv4PDqX:-499146009 -->

---

<sup>**[flocker/node/testtools.py, line 242 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnDICX6If6RXNH--r1-242)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/testtools.py#L242)):</sup>
I actually added this because we need to be sure it is AMP serializable, so I want to explicitly verify that it is a tuple (lists would also be allowed I suppose). Arbitrary iterables that AMP cannot serialize will pass the next test but not this one.
<!-- Reviewable comment -K0DYY_zr6Z3zu6xoI-W:-1391939492 -->

---

<sup>**[flocker/node/testtools.py, line 312 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnDICX6If6RXNH--r1-312)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/testtools.py#L312)):</sup>
Realized we can't actually have a default at this layer - this function is called for different types of IDeployers that have different types of ILocalStates - making this a  mandatory argument and adding defaults at the next layer up.
<!-- Reviewable comment -K0DostWUNnGOnnd56Hp:-481132397 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/clusterhq/flocker/2046)
<!-- Sent from Reviewable.io -->

itamarst commented on flocker/node/agents/test/test_blockdevice.py:
```diff
@@ -698,6 +752,186 @@ def add_application_with_volume(node_state):
+    return volumes
+
+
+# Constants for determining volumes in calculate_changes tests. See
```
I believe you need to have these as class attributes (http://twistedmatrix.com/documents/current/core/howto/constants.html)

itamarst commented on flocker/node/agents/test/test_blockdevice.py:
```diff
@@ -698,6 +752,186 @@ def add_application_with_volume(node_state):
+        volumes = _infer_volumes_for_test(
+            block_device_deployer.node_uuid, block_device_deployer.hostname,
+            compute_instance_id, cluster_state)
+    elif volumes is _GET_VOLUMES_FROM_API:
```
this case looks like it's used only in one test, so maybe just do the action manually there and simplify this function.

itamarst commented on flocker/node/agents/test/test_blockdevice.py:
```diff
@@ -698,6 +752,186 @@ def add_application_with_volume(node_state):
                                   mountpoint=FilePath(b"/data")))})
 
 
+_BLOCKDEVICE_PREFIX = "blockdevice-"
```
This is necessary because tests manually construct state. Arguably that's a mistake in test design since it's brittle. Probably not worth spending any more time on this so let's stick to this approach, but at least write comment explaining motivation and suggestion for improvement (always make state using API as in realistic usage?)

itamarst commented on flocker/node/agents/test/test_blockdevice.py:
```diff
@@ -698,6 +752,186 @@ def add_application_with_volume(node_state):
+_BLOCKDEVICE_PREFIX = "blockdevice-"
+
+
+def _get_dataset_id_for_test_blockdevice_id(test_blockdevice_id):
```
This function seems to be unused.

itamarst commented on flocker/node/agents/test/test_blockdevice.py:
```diff
@@ -698,6 +752,186 @@ def add_application_with_volume(node_state):
+    return "".join([_BLOCKDEVICE_PREFIX, dataset_id])
+
+
+def _create_test_blockdevice_volume_for_dataset_id(dataset_id,
```
I would just inline this into  next function since it's only used there, and then you don't need to document it.

```
def f():
    def g(): ...
    x = g()
```

itamarst commented on flocker/node/agents/test/test_blockdevice.py:
```diff
@@ -698,6 +752,186 @@ def add_application_with_volume(node_state):
+    return test_blockdevice_id[len(_BLOCKDEVICE_PREFIX):]
+
+
+def _create_test_blockdevice_id_for_dataset_id(dataset_id):
```
Doesn't need to be separate function. Also can probably just do `_BLOCKDEVICE_PREFIX + dataset_id`.

itamarst replied:
Or at least "inline" it (wrong word I guess), see next comment.

itamarst commented on flocker/node/agents/test/test_blockdevice.py:
```diff
@@ -698,6 +752,186 @@ def add_application_with_volume(node_state):
+    local_volumes = [
+        _create_test_blockdevice_volume_for_dataset_id(
+            dataset_id, attached_to=compute_instance_id)
+        for dataset_id in (node_state.manifestations or {}).keys()]
```
Unless this can be `None` you don't need `or {}`.

itamarst replied:
Nor do you need `keys()`, iterating over dictionary gives its keys.

itamarst commented:
OK, a bunch of minor comments to address, then merge when buildbot is green.

sarum90 commented:
Closed out a bunch of comments (I think? still getting used to reviewable and figuring out how I want to use it).

Thanks for the review! Addressed all comments except 1 which appeared to just be a minor one for what I think is just stylistic differences.
<!-- Reviewable comment -K15Mz0YGZV3V3gY_xoS:1894970441 -->

---

Review status: 0 of 10 files reviewed at latest revision, 20 unresolved discussions, all commit checks successful.

---

<sup>**[flocker/node/_deploy.py, line 67 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnCwSm_DOGOIVNO-r1-67)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/_deploy.py#L67)):</sup>
Done.
<!-- Reviewable comment -K15IAhx4cF2X0uTyCYK:-499146009 -->

---

<sup>**[flocker/node/_deploy.py, line 88 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnCwSm_DOGOIVNO-r1-88)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/_deploy.py#L88)):</sup>
Done.
<!-- Reviewable comment -K15ICK7iLQYdpaM5Vft:-499146009 -->

---

<sup>**[flocker/node/_deploy.py, line 99 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnCwSm_DOGOIVNO-r1-99)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/_deploy.py#L99)):</sup>
Done.
<!-- Reviewable comment -K15II-0clXIvAlTVWpd:-499146009 -->

---

<sup>**[flocker/node/_deploy.py, line 129 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnCwSm_DOGOIVNO-r1-129)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/_deploy.py#L129)):</sup>
Done.
<!-- Reviewable comment -K15IGyrxkx1PFnCB85d:-499146009 -->

---

<sup>**[flocker/node/agents/blockdevice.py, line 1584 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnCwSm_DOGOIVNQ-r1-1584)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/agents/blockdevice.py#L1584)):</sup>
Done.
<!-- Reviewable comment -K15IEoahPqh7gprWCX8:-499146009 -->

---

<sup>**[flocker/node/agents/test/test_blockdevice.py, line 745 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnCwSm_DOGOIVNR-r1-745)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/agents/test/test_blockdevice.py#L745)):</sup>
Done.
<!-- Reviewable comment -K15HJcEIDZM0x1PrJ_3:-499146009 -->

---

<sup>**[flocker/node/agents/test/test_blockdevice.py, line 755 \[r3\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnCwSm_DOGOIVNR-r3-755)** ([raw file](https://github.com/clusterhq/flocker/blob/3bd25b017c871b426fec90ecffb325e40cebe95f/flocker/node/agents/test/test_blockdevice.py#L755)):</sup>
After "inlining" this comment was more naturally added to the docstring for _infer_volumes_for_test. Extended that comment to describe its own brittleness and encourage tests not to use it.
<!-- Reviewable comment -K157jhtAyfYQUAG2Pv0:281949231 -->

---

<sup>**[flocker/node/agents/test/test_blockdevice.py, line 758 \[r3\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnCwSm_DOGOIVNR-r3-758)** ([raw file](https://github.com/clusterhq/flocker/blob/3bd25b017c871b426fec90ecffb325e40cebe95f/flocker/node/agents/test/test_blockdevice.py#L758)):</sup>
Removed - Done. I might have to add it back soon though - if and when that happens it'll be obvious why and we can defer conversations 'til that time.
<!-- Reviewable comment -K154F_9vMkQQh68kDtM:1894341484 -->

---

<sup>**[flocker/node/agents/test/test_blockdevice.py, line 773 \[r3\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnCwSm_DOGOIVNR-r3-773)** ([raw file](https://github.com/clusterhq/flocker/blob/3bd25b017c871b426fec90ecffb325e40cebe95f/flocker/node/agents/test/test_blockdevice.py#L773)):</sup>
Done.
<!-- Reviewable comment -K15ANISfTCa11FYQ3Kp:-499146009 -->

---

<sup>**[flocker/node/agents/test/test_blockdevice.py, line 790 \[r3\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnCwSm_DOGOIVNR-r3-790)** ([raw file](https://github.com/clusterhq/flocker/blob/3bd25b017c871b426fec90ecffb325e40cebe95f/flocker/node/agents/test/test_blockdevice.py#L790)):</sup>
Done.
<!-- Reviewable comment -K15APh06hWLJiE7tXnY:-499146009 -->

---

<sup>**[flocker/node/agents/test/test_blockdevice.py, line 849 \[r3\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnCwSm_DOGOIVNR-r3-849)** ([raw file](https://github.com/clusterhq/flocker/blob/3bd25b017c871b426fec90ecffb325e40cebe95f/flocker/node/agents/test/test_blockdevice.py#L849)):</sup>
Yea.. some tests actually do pass None in (BlockDeviceDeployerCreationCalculateChangesTests.test_no_devices_no_local_datasets for instance). And in general it is valid for node_state.manifestations to be None (indicates a state that doesn't know about manifestations).

removed the ".keys()"
<!-- Reviewable comment -K15CLQIcH46LxdX901G:544890398 -->

---

<sup>**[flocker/node/agents/test/test_blockdevice.py, line 877 \[r3\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnCwSm_DOGOIVNR-r3-877)** ([raw file](https://github.com/clusterhq/flocker/blob/3bd25b017c871b426fec90ecffb325e40cebe95f/flocker/node/agents/test/test_blockdevice.py#L877)):</sup>
Thanks for catching that! fixed.
<!-- Reviewable comment -K15FKsHmN1XQG1vrwX9:-1587340944 -->

---

<sup>**[flocker/node/agents/test/test_blockdevice.py, line 923 \[r3\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnCwSm_DOGOIVNR-r3-923)** ([raw file](https://github.com/clusterhq/flocker/blob/3bd25b017c871b426fec90ecffb325e40cebe95f/flocker/node/agents/test/test_blockdevice.py#L923)):</sup>
Acknowledged. I sort of like using defaults to imply best practices. I think it's best if the caller just uses the api to construct the volumes, and then doesn't have to do anything at this layer. Hopefully these two lines and the extra constant don't add that much complexity.
<!-- Reviewable comment -K15HLIHhLW5RcXzJdUy:-1921104981 -->

---

<sup>**[flocker/node/functional/test_deploy.py, line 52 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnCwSm_DOGOIVNS-r1-52)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/functional/test_deploy.py#L52)):</sup>
Done.
<!-- Reviewable comment -K15Hz6SDRjbwkQydyQN:-499146009 -->

---

<sup>**[flocker/node/functional/test_deploy.py, line 60 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnCwSm_DOGOIVNS-r1-60)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/functional/test_deploy.py#L60)):</sup>
Done.
<!-- Reviewable comment -K15IQ-ABcwBpFxo_gfR:-499146009 -->

---

<sup>**[flocker/node/test/test_deploy.py, line 193 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnDICX6If6RXNGz-r1-193)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/test/test_deploy.py#L193)):</sup>
Done.
<!-- Reviewable comment -K15IR0K3998HcHlEYw6:-499146009 -->

---

<sup>**[flocker/node/test/test_deploy.py, line 714 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnDICX6If6RXNGz-r1-714)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/test/test_deploy.py#L714)):</sup>
Done.
<!-- Reviewable comment -K15IS220sR8SvfYimav:-499146009 -->

---

<sup>**[flocker/node/test/test_deploy.py, line 1458 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnDICX6If6RXNGz-r1-1458)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/test/test_deploy.py#L1458)):</sup>
Done.
<!-- Reviewable comment -K15ITAa1tp_73sNFo-y:-499146009 -->

---

<sup>**[flocker/node/testtools.py, line 154 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnDICX6If6RXNH--r1-154)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/testtools.py#L154)):</sup>
Done.
<!-- Reviewable comment -K15ITlkEFyHHpolpYFJ:-499146009 -->

---

<sup>**[flocker/node/testtools.py, line 312 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2046#-K08nCnDICX6If6RXNH--r1-312)** ([raw file](https://github.com/clusterhq/flocker/blob/ec8dda4a394112cafc0989623c36e9252b806d9f/flocker/node/testtools.py#L312)):</sup>
Done.
<!-- Reviewable comment -K15IUUHGVDFvyk5y7U_:-499146009 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/clusterhq/flocker/2046)
<!-- Sent from Reviewable.io -->

sarum90 commented:
I just rebased this onto master to get Tom's fix in the branch. Re-running a full buildbot run presently with my fingers crossed.
